examples/prom/de_parametric_maxwell_greedy.cpp

```diff
@@ -0,0 +1,953 @@
+int precision = 8;
+double coef = 1.0;
+
+// Greedy algorithm variables
+bool build_database = false;
+bool use_database = false;
+double greedy_param_space_min = 1.0;
+double greedy_param_space_max = 1.2;
+int greedy_param_space_size = 5;
+double greedy_relative_error_tol = 0.01;
+int greedy_subset_size = 2;
+int greedy_convergence_subset_size = 3;
+double local_rom_freq = 0.0;
+double curr_freq = 0.0;
+vector<string> basisIdentifiers;
+CAROM::GreedySampler* greedy_sampler = NULL;
+bool calc_rel_error;
+std::string curr_basis_identifier;
+bool calc_err_indicator = false;
+
+// DE variables
+bool de = false;
+double de_min_freq= -DBL_MAX;
+double de_max_freq = DBL_MAX;
+double target_freq = freq;
+double de_F = 0.8;
+double de_CR = 0.9;
+int de_PS = 50;
+int de_min_iter = 10;
+int de_max_iter = 100;
+double de_ct = 0.001;
+#ifdef MFEM_USE_AMGX
+bool useAmgX = false;
+#endif
+
+// Define the process for solving Maxwell equation. We will repeatedly call this function later
+double simulation()
+{
+    if (myid == 0) cout << "Running loop at frequency: " << freq << endl;
+
+    kappa = freq * M_PI;
+
+    //    Enable hardware devices such as GPUs, and programming models such as
+    //    CUDA, OCCA, RAJA and OpenMP based on command line options.
+    Device device(device_config);
+    if (myid == 0 || !de) {
+        device.Print();
+    }
+
+    // 6. Read the (serial) mesh from the given mesh file on all processors.  We
+    //    can handle triangular, quadrilateral, tetrahedral, hexahedral, surface
+    //    and volume meshes with the same code.
+    Mesh *mesh = new Mesh(mesh_file, 1, 1);
+    dim = mesh->Dimension();
+    int sdim = mesh->SpaceDimension();
+
+    // 7. Refine the serial mesh on all processors to increase the resolution. In
+    //    this example we do 'ref_levels' of uniform refinement. We choose
+    //    'ref_levels' to be the largest number that gives a final mesh with no
+    //    more than 1,000 elements.
+    {
+        int ref_levels = (int)floor(log(1000./mesh->GetNE())/log(2.)/dim);
+        for (int l = 0; l < ref_levels; l++)
+        {
+            mesh->UniformRefinement();
+        }
+    }
+
+    // 8. Define a parallel mesh by a partitioning of the serial mesh. Refine
+    //    this mesh further in parallel to increase the resolution. Once the
+    //    parallel mesh is defined, the serial mesh can be deleted.
+    ParMesh *pmesh = new ParMesh(MPI_COMM_WORLD, *mesh);
+    delete mesh;
+    {
+        int par_ref_levels = 2;
+        for (int l = 0; l < par_ref_levels; l++)
+        {
+            pmesh->UniformRefinement();
+        }
+    }
+
+    // 9. Define a parallel finite element space on the parallel mesh. Here we
+    //    use the Nedelec finite elements of the specified order.
+    FiniteElementCollection *fec = new ND_FECollection(order, dim);
+    ParFiniteElementSpace *fespace = new ParFiniteElementSpace(pmesh, fec);
+    HYPRE_BigInt size = fespace->GlobalTrueVSize();
+    if (!de)
```

Comment:
Minor improvement: `if (!de && myid == 0)`

Reply:
Good point, I fix the few instances where I wrote things this way.